### PR TITLE
first commit

### PR DESCRIPTION
**Pull request by golnazads**

### CODE REVIEW

File: referencesrv/views.py [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+
+bp = Blueprint('reference_service', __name__)
+
+
+def text_parser(reference):
+    """
+
+    :return:
+    """
+    if not hasattr(text_parser, "crf"):
```
Comment (romanchyla) [inline referencesrv/views.py]: is this setting and attribute of itself? (function)

that seems funny

File: referencesrv/tests/unittests/test_referencesrv_parser.py [hunk 2]
```diff
@@ -0,0 +1,220 @@
+import sys, os
+project_home = os.path.abspath(os.path.join(os.path.dirname(__file__), '../../../'))
```
[the remark this answers is not in the document]
Comment (golnazads) [reply referencesrv/tests/unittests/test_referencesrv_parser.py]: Let me remove it, run couple of examples and see if we come across it. Just to document it.

File: referencesrv/parser/crf.py [hunk 3]
```diff
@@ -0,0 +1,1255 @@
...
+        :param index: the position of the word in the set
+        :param ref_label: labels for ref_data available during training only
+        :param noun_phrase
+        :return:
+        """
+        return [
+            len(ref_data[index]),                                                                                       # length of element
+            len(ref_data[index-1]) if index > 0 else 0,                                                                 # length of previous element if any
+            len(ref_data[index+1]) if index < len(ref_data) - 1 else 0,                                                 # length of next element if any
+            int(bool(re.match('^([A-Z]+)$', ref_data[index]))),                                                         # is element all capital
```
[the remark this answers is not in the document]
Comment (golnazads) [reply referencesrv/parser/crf.py]: YES. great point. thank you so much.

File: referencesrv/tests/unittests/test_referencesrv_parser.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (golnazads) [reply referencesrv/tests/unittests/test_referencesrv_parser.py]: mystery! not getting any errors. so removed it. thank you.

File: referencesrv/resolver/pytrigdict.py [hunk 4]
```diff
@@ -0,0 +1,221 @@
+"""
+A trigdict implements a fuzzy dictionary matching keys with
+Levenshtein weights.  For efficiency reasons, before computing
+the distances, we go through a trigram index to find candates.
+
+NOTE: over the original C-based implementation, the actual scores
```
[the remark this answers is not in the document]
Comment (golnazads) [reply referencesrv/resolver/pytrigdict.py]: lol. but you know that is not me, all Marcus.
in reply to the c-implementation, this is the imported code to python so no need for the c-implementation.

File: referencesrv/views.py (hunk 1 above)
Comment (golnazads) [reply referencesrv/views.py]: yes for the function. and that is how it grantees there is only one time training.

File: referencesrv/parser/crf.py [hunk 5]
```diff
@@ -0,0 +1,1255 @@
...
+            numeric_labels.append(np.array(numeric_label))
+
+            # get the numeric features for the original presentation of word and insert at index of label
+            feature = []
+            for idx in range(len(word)):
+                feature.append(self.get_data_features(word, idx, label))
+            features.append(np.array(feature))
+        return features, numeric_labels, label_code
+
+    def get_ready(self):
```
[the remark this answers is not in the document]
Comment (golnazads) [reply referencesrv/parser/crf.py]: see the comment below. so the function attributes prevents multiple training.

**Pull request merged**